Add concurrency model with uv loops

IsabelParedes opened the pull request.

IsabelParedes committed:
Add new uv shell

IsabelParedes committed:
Add uvw dependency

IsabelParedes committed:
Copy split class

IsabelParedes committed:
Replace virtual fun

IsabelParedes committed:
Copy control and shell for uv server

IsabelParedes committed:
Add make fun for uv server

IsabelParedes committed:
Replace pool with uvw loop

IsabelParedes committed:
Rebase

Resolve rebase changes

IsabelParedes committed:
Add uvw to environment file

IsabelParedes committed:
Update header paths

IsabelParedes committed:
Update make uv fun

IsabelParedes committed:
Fix header guards

IsabelParedes committed:
Register callback

IsabelParedes committed:
Replace poll event

IsabelParedes committed:
Add controller poll

IsabelParedes committed:
Use uvw v3.3.0

IsabelParedes committed:
Fix uvw linkage

uvw symbols were undefined when testing with xeus-python

IsabelParedes committed:
Test polls

IsabelParedes committed:
Polling

IsabelParedes committed:
Fix json error

IsabelParedes committed:
Debug kernel status

IsabelParedes committed:
Get loop pointer

IsabelParedes committed:
Use loop pointer

IsabelParedes committed:
Run loop

IsabelParedes committed:
Debug loop run

IsabelParedes committed:
Save poll resources

IsabelParedes committed:
Refactor

IsabelParedes committed:
Unoverload uv server maker

IsabelParedes committed:
Add hook base

Fix linking issue

IsabelParedes committed:
Clean up

IsabelParedes commented:
Windows tests were failing because they were using the C++11 standard.  https://github.com/jupyter-xeus/xeus-zmq/pull/38

IsabelParedes committed:
Add default uvw loop

IsabelParedes committed:
Clean

JohanMabille commented on include/xeus-zmq/xhook_base.hpp:
```diff
@@ -0,0 +1,37 @@
+        virtual void pre_hook() = 0;
+        virtual void post_hook() = 0;
+
+        virtual void run(std::shared_ptr<uvw::loop> loop)
```
Public methods should not be virtual, especially in public headers. Prefer public non virtual methods (except for the destructor) that internally call private virtual methods. This gives more flexibility in the implementation while offering a better API stability. See [here](http://www.gotw.ca/publications/mill18.htm) for more detail.

JohanMabille commented on include/xeus-zmq/xhook_base.hpp:
```diff
@@ -0,0 +1,37 @@
+        virtual void run(std::shared_ptr<uvw::loop> loop)
+        {
+            loop->run();
+        };
```
As much as possible, implementation in public header should be avoided.

JohanMabille commented on include/xeus-zmq/xhook_base.hpp:
```diff
@@ -0,0 +1,37 @@
+
+namespace xeus
+{
+    class xhook_base
```
This should be exported with the `XEUS_ZMQ_API` macro, otherwise the symbol won't be found on Windows.

JohanMabille commented on include/xeus-zmq/xhook_base.hpp:
```diff
@@ -0,0 +1,37 @@
+    {
+    public:
+
+        virtual ~xhook_base() = default;
```
This class should implement proper entity semantics:
- consider declaring a default constructor in the protected section
- consider deleting the copy and move semantics (i.e. constructor + assignment operator) in the public section.

JohanMabille commented on include/xeus-zmq/xserver_zmq.hpp:
```diff
@@ -10,11 +10,19 @@
 
+#ifndef UVW_AS_LIB
+#define UVW_AS_LIB
+#include <uvw.hpp>
```
Shouldn't we include this header even if `UVW_AS_LIB` is not defined? Otherwise, a short comment explaining how uvw primitives are found could be helpful.

IsabelParedes replied:
I moved these `UVW_AS_LIB` directives to the CMakeLists file as `target_compile_definitions()` instead, to make it cleaner and to include uvw every time it's needed without compilation errors.

JohanMabille commented on include/xeus-zmq/xserver_zmq.hpp:
```diff
@@ -63,6 +71,15 @@ namespace xeus
                                                      nl::json::error_handler_t eh = nl::json::error_handler_t::strict);
+
+    XEUS_ZMQ_API
+    std::unique_ptr<xserver> make_xserver_uv_shell_main(
```
I think adding a comment showing how this function should be used could be helpful. The `xeus::xkernel` constructor accepts a function that takes 3 arguments only, so on would probably have to use a lamba capturing the loop and the hook before calling this function.

JohanMabille commented on src/xserver_zmq_split.cpp:
```diff
@@ -10,6 +10,11 @@
 
+#ifndef UVW_AS_LIB
+#define UVW_AS_LIB
+#include <uvw.hpp>
```
Same comment as above regarind the conditional include.

JohanMabille commented on src/xserver_zmq_split.cpp:
```diff
@@ -33,7 +40,32 @@ namespace xeus
+                                     config.m_transport, config.m_ip, config.m_iopub_port))
+        , p_shell(new xshell_uv(context, config.m_transport, config.m_ip,
+                                config.m_shell_port, config.m_stdin_port, this,
+                                loop_ptr, std::move(hook)))
```
This should be factorized with a private delegating constructor. The public constructors should initialize the right shell object, and forward it (and everything else) to the delegating constructor.

IsabelParedes replied:
I updated this, but I'm not 100% if this is what you meant. Please let me know. Thanks!

JohanMabille commented on src/xshell_base.cpp:
```diff
@@ -39,72 +39,30 @@ namespace xeus
     }
 
-    void xshell::run()
+    void xshell_base::run()
```
It looks that this methods should forward the call to a pure virtual implementation method.

JohanMabille commented on src/xshell_base.hpp:
```diff
@@ -21,25 +21,25 @@ namespace xeus
                xserver_zmq_split* server);
 
-        ~xshell();
+        ~xshell_base();
```
This constructor should be virtual now.

JohanMabille commented on src/xshell_base.hpp:
```diff
@@ -21,25 +21,25 @@ namespace xeus
         using listener = std::function<void(xmessage)>;
 
-        xshell(zmq::context_t& context,
+        xshell_base(zmq::context_t& context,
```
This class should implement proper entity semantics:
- the constructor should be moved to a protected section
- copy and move semantics should be deleted in the public section.

JohanMabille commented on src/xshell_base.hpp:
```diff
@@ -21,25 +21,25 @@ namespace xeus
         std::string get_stdin_port() const;
 
-        void run();
+        virtual void run();
```
As for the hook class, this should be non virtual and forward the call to a private virtual method.

JohanMabille commented on src/xshell_default.cpp:
```diff
@@ -0,0 +1,77 @@
+
+    void xshell_default::run()
+        {
+            zmq::pollitem_t items[] = {
```
Nitpicking: alignment

JohanMabille commented on src/xshell_default.hpp:
```diff
@@ -0,0 +1,36 @@
+                       const std::string& stdin_port,
+                       xserver_zmq_split* server);
+
+        ~xshell_default();
```
Should be virtual

JohanMabille commented on src/xshell_default.cpp:
```diff
@@ -0,0 +1,77 @@
+    {
+    }
+
+    xshell_default::~xshell_default()
```
This can be defaulted in the header.

JohanMabille commented on src/xshell_uv.cpp:
```diff
@@ -0,0 +1,125 @@
+        create_polls();
+    }
+
+    xshell_uv::~xshell_uv()
```
This can be defaulted in the header.

JohanMabille commented on src/xshell_uv.cpp:
```diff
@@ -0,0 +1,125 @@
+        if (p_hook)
+        {
+            p_hook->run(p_loop);
+        } else {
```
Nitpicking: opening curly brace should go on a new line.

JohanMabille commented on src/xshell_uv.hpp:
```diff
@@ -0,0 +1,56 @@
+                       std::shared_ptr<uvw::loop> loop_ptr,
+                       std::unique_ptr<xhook_base> hook);
+
+        ~xshell_uv();
```
This should be virtual.

IsabelParedes replied:
Should these derived classes, `xshell_uv` and `xshell_default` have virtual destructors, or should they *override* the virtual destructor of the base class?
```cpp
virtual ~xshell_base() = default;
```

JohanMabille commented on test/xmock_interpreter.cpp:
```diff
@@ -8,6 +8,11 @@
+#ifndef UVW_AS_LIB
+#define UVW_AS_LIB
+#include <uvw.hpp>
+#endif
```
Why do we need this include here?

JohanMabille reviewed (changes requested):
Thanks a lot for implementing this! Beyond the comments I feft, I wonder if we should type erase the uvw pointer in the `make_xserver_uv_main_shell` function, so that we can abstract it and remove the dependency to uvw in xeus-zmq in a future iteration.

IsabelParedes committed:
Update xhook_base

This commit privatizes virtual methods, exports the class with a macro,
removes the implementation from the public header, and implements proper
entity semantics.

IsabelParedes committed:
Implement entity semantics

IsabelParedes committed:
Fix shell classes

IsabelParedes committed:
Fix alignment

IsabelParedes committed:
Fix rebase conflict

IsabelParedes committed:
Restore tests

IsabelParedes committed:
Make shell destructors virtual

IsabelParedes committed:
Add delegating constructor

IsabelParedes committed:
Fix alignment

IsabelParedes committed:
Clean up uvw preprocessor directives

IsabelParedes commented:
> I wonder if we should type erase the uvw pointer in the `make_xserver_uv_main_shell` function, so that we can abstract it and remove the dependency to uvw in xeus-zmq in a future iteration.

Good idea. I'll look into this

IsabelParedes committed:
Add comment to create uv xserver

IsabelParedes committed:
Stop loop

IsabelParedes commented:
Closing this in favor of https://github.com/jupyter-xeus/xeus-uv
